# Urchin local adaptation
This repo was made to cleanly demonstrate how I got from raw NGS data to different sets of loci putatively under positive selection using Fst outlier, Bayenv and reduced nucleotide diversity measures.

## Step 1: From reads to bams

[Code for Step 1](https://github.com/Cpetak/urchin_adaptation/blob/main/Step1.md)

## Step 2: from bams to genotype likelihoods and PCA
[Code for Step 2](https://github.com/Cpetak/urchin_adaptation/blob/main/Step2.md)

## Step 3: Getting per-site Fst values - global
[Code for Step 3](https://github.com/Cpetak/urchin_adaptation/blob/main/Step3.md)


## Step 4: Running LFMM

[Code for Step 4](https://github.com/Cpetak/urchin_adaptation/blob/main/Step4.md)

## Step 5: PCAngsd

[Code for Step 5](https://github.com/Cpetak/urchin_adaptation/blob/main/Step5.md)

## Step 6: Getting per-site Fst values - pair-wise

[Code for Step 6](https://github.com/Cpetak/urchin_adaptation/blob/main/Step6.md)

## Step 7: Getting pair-wise global Fst values

[Code for Step 7](https://github.com/Cpetak/urchin_adaptation/blob/main/Step7.md)

## Step 8: Per-site nucleotide diversity

[Code for Step 8](https://github.com/Cpetak/urchin_adaptation/blob/main/Step8.md)

